saleor/graphql/account/mutations/jwt.py

```diff
@@ -0,0 +1,258 @@
+        user = cls.get_user(info, data)
+        access_token = create_access_token(user)
+        csrf_token = _get_new_csrf_token()
+        refresh_token = create_refresh_token(user, {"csrfToken": csrf_token})
+        info.context.refresh_token = refresh_token
+        info.context._cached_user = user
+        user.last_login = timezone.now()
+        user.save(update_fields=["last_login"])
+        return cls(
+            errors=[],
+            user=user,
+            token=access_token,
+            refresh_token=refresh_token,
+            csrf_token=csrf_token,
+        )
+
+
+class RefreshToken(BaseMutation):
+    """Mutation that refresh user token and returns token and user data."""
+
+    token = graphene.String(description="JWT token, required to authenticate.")
+    user = graphene.Field(User, description="A user instance.")
+
+    class Arguments:
+        refresh_token = graphene.String(required=False, description="Refresh token.")
+        csrf_token = graphene.String(
+            required=True, description="CSRF token required to refresh token."
+        )
+
+    class Meta:
+        description = (
+            "Refresh JWT token. Mutation tries to take refreshToken from "
+            f"the http-only cookie {JWT_REFRESH_TOKEN_COOKIE_NAME}. If it "
+            "fails it will try to take refreshToken provided as an argument."
+        )
+        error_type_class = AccountError
+        error_type_field = "account_errors"
+
+    @classmethod
+    def get_refresh_token_payload(cls, refresh_token):
+        try:
+            payload = get_payload(refresh_token)
+        except ValidationError as e:
+            raise ValidationError({"refreshToken": e})
+        return payload
+
+    @classmethod
+    def get_refresh_token(cls, info, data):
+        request = info.context
+        refresh_token = request.COOKIES.get(JWT_REFRESH_TOKEN_COOKIE_NAME, None)
+        refresh_token = data.get("refresh_token") or refresh_token
+        return refresh_token
+
+    @classmethod
+    def clean_refresh_token(cls, refresh_token):
+        if not refresh_token:
+            raise ValidationError(
+                {
+                    "refreshToken": ValidationError(
+                        "Missing refreshToken",
+                        code=AccountErrorCode.JWT_MISSING_TOKEN.value,
+                    )
+                }
+            )
+        payload = cls.get_refresh_token_payload(refresh_token)
+        if payload["type"] != JWT_REFRESH_TYPE:
+            raise ValidationError(
+                {
+                    "refreshToken": ValidationError(
+                        "Incorrect refreshToken",
+                        code=AccountErrorCode.JWT_INVALID_TOKEN.value,
+                    )
+                }
+            )
+        return payload
+
+    @classmethod
+    def clean_csrf_token(cls, csrf_token, payload):
+        is_valid = _compare_salted_tokens(csrf_token, payload["csrfToken"])
```

Comment:
I'm wondering whether we could limit the CSRF requirement to cases where the refresh token is read from a cookie?